web/src/zform.js

```diff
@@ -1,43 +1,68 @@
-                return schema.check_record(field_name, val, {
-                    short_name: schema.check_string,
-                    long_name: schema.check_string,
-                    reply: schema.check_string,
-                });
-            }
-
-            function check_choices(field_name, val) {
-                return schema.check_array(field_name, val, check_choice_item);
-            }
-
-            return schema.check_record("zform data", data, {
-                heading: schema.check_string,
-                choices: check_choices,
-            });
-        }
-
-        if (data.type === "choices") {
-            return check_choice_data(data);
-        }
-
-        return "unknown zform type: " + data.type;
+    if (data.type !== "choices") {
+        blueslip.warn(`unknown zform type: ${data.type}`);
+        return false;
     }
 
-    const msg = check(data);
-
-    if (msg) {
-        blueslip.warn(msg);
+    const zFormSchema = z.object(
+        {
+            choices: z.array(
+                z.object(
+                    {
+                        long_name: z.string({
+                            invalid_type_error: "long_name is not a string",
+                            required_error: "long_name is not provided",
+                        }),
+                        reply: z.string({
+                            invalid_type_error: "reply is not a string",
+                            required_error: "reply is not provided",
+                        }),
+                        short_name: z.string({
+                            invalid_type_error: "short_name is not a string",
+                            required_error: "short_name is not provided",
+                        }),
+                    },
+                    {
+                        errorMap(issue, ctx) {
+                            if (issue.code === "invalid_type") {
+                                return {
+                                    message: "element of choices array is not a record",
+                                };
+                            }
+                            return {message: ctx.defaultError};
+                        },
+                    },
+                ),
+                {
+                    invalid_type_error: "choices is not an array",
+                    required_error: "choices is not provided",
+                },
+            ),
+            heading: z.string({
+                invalid_type_error: "heading is not a string",
+                required_error: "heading is not provided",
+            }),
+        },
+        {
+            invalid_type_error: `zform data is not a record`,
+            required_error: "zform data is not provided",
+        },
+    );
+
+    const parseResult = zFormSchema.safeParse(data, {
+        path: ["zform data"],
+    });
+
+    if (!parseResult.success) {
+        blueslip.warn("invalid zform data", parseResult.error.issues);
         return false;
```

Comment:
Zod is designed to return the parsed object—which may be different from the original object—not just a boolean `true` or `false`. That is better for type safety, so we should update our API to match.

Reply:
@andersk You mean I should change the code to make use of the object returned by zod after parsing?

Reply:
@andersk take a look